src/hAPI/README.md

```diff
@@ -1,5 +1,54 @@
+ 
+The worker is instantiated in the ```hapi-utils.ts``` script by sending a preliminary message containing the  data points for the segments and objects contained in the image. The normalized coordinates returned from the handler are mapped into the 2DIY's frame of reference and then grouped into specific data types (segments, or objects) which are then utilized in functions called from the haptic loop to trace contours or move to certain positions when necessary.
+
+The worker is currently set up to infinitely loop, refreshing at a rate of 1 ms. However, stability issues occur when ensuring that this is maintained at all times. 
+
+Within a single haptic loop, the worker is responsible for reading the current position of the end effector, communicating this position information along with state tracking variables to the main script, and finally calculating the required forces which are transmitted to the 2DIY.
+
+The ```moveToPos()``` function is responsible for setting the forces based on the next position, calculated using the difference between the current end-effector position and next in the index of the current subsegment or object. It utilizes a PID filter and 5-way moving average filter.
+
+For grouped objects, a convex hull is used to generate an array 'connecting' the centroid of each object within the group. Interpolation is then used to generate a much larger array of points using the ```upsample()``` function, which allows for the 2DIY to smoothly trace a path connecting the object centroids as if they formed a segment.
+
+## Visuals
+
+The visuals are generated in the ```hapi-utils.ts``` file. An HTML canvas is created under the dropdown for the haptic rendering using the ```createCanvas()``` function. This canvas displays the photo along with a visual avatar of the end effector's instantaneous position and the outline of any segment or object being rendered. 
+
+The [requestAnimationFrame(draw)](https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame) function is responsible for drawing graphics on the canvas, including the photograph itself. The ```updateAnimation()``` called by the ```draw()``` callback function takes in the current position of the 2DIY's end-effector and information about the current segment and object, obtained from the web worker's haptic loop. Because the callback function is called 60 times a second, it is advisable to only draw immediately necessary information, such as the 2DIY avatar and outline of the current subsegment or object being drawn only.
+
+## hAPI TypeScript Library 
+
+The [hAPI library](https://github.com/Shared-Reality-Lab/IMAGE-browser/tree/main/src/hAPI/libraries) used in the project is a TypeScript port of the [Java hAPI library](https://gitlab.com/Haply/hAPI) originally authored by Oliver Anthony, [Haply Robotics](https://haply.co/). ```Vector.ts``` is an additional helper file for vector calculations.
```

Comment:
IIRC Vector.ts is heavily adapted from another library's Vector.js. If this is the case, please provide attribution to the original library.

Reply:
Added.